src/main/scala/com/fulcrumgenomics/alignment/Aligner.scala

```diff
@@ -75,14 +75,30 @@ object Aligner {
   /** The minimum score allowed to start an alignment.  This prevents underflow. */
   val MinStartScore: Int = Int.MinValue / 2
 
+  /** Specialized matrix for storing integers that is markedly faster than using LinearMatrix[Int]. */
+  final class SimpleIntMatrix(val x: Int, val y: Int) extends Matrix[Int] {
+    private val data = new Array[Array[Int]](x)
+    forloop (from=0, until=x) { i => data(i) = new Array[Int](y) }
+
+    override def apply(i: Int, j: Int): Int = data(i)(j)
+    override def update(i: Int, j: Int, value: Int): Unit = data(i)(j) = value
+    override def dimensions: (Int, Int) = (x, y)
+  }
+
   object AlignmentMatrix {
     def apply(direction: Direction, queryLength: Int, targetLength: Int): AlignmentMatrix = {
-      AlignmentMatrix(direction=direction, scoring=Matrix[Int](queryLength+1, targetLength+1), trace=Matrix[Direction](queryLength+1, targetLength+1))
+      AlignmentMatrix(
+        direction = direction,
+        scoring   = new SimpleIntMatrix(queryLength+1, targetLength+1),
+        trace     = new SimpleIntMatrix(queryLength+1, targetLength+1))
     }
   }
 
   /** A single alignment matrix for a given [[Direction]] storing both the scoring and traceback matrices produce by the aligner. */
-  case class AlignmentMatrix(direction: Direction, scoring: Matrix[Int], trace: Matrix[Direction])
+  case class AlignmentMatrix(direction: Direction, scoring: Matrix[Int], trace: Matrix[Direction]) {
+    val queryLength: Int  = scoring.dimensions._1 - 1
+    val targetLength: Int = scoring.dimensions._2 - 1
```

Comment:
It would, but those vals are not defined in `Matrix`.  I just extended the `Matrix` trait to add them though since they're useful an both implementations have them.